pkg/lib/bundle/generate.go

```diff
@@ -46,30 +50,27 @@ type AnnotationMetadata struct {
 // @channels: The list of channels that bundle image belongs to
 // @channelDefault: The default channel for the bundle image
 // @overwrite: Boolean flag to enable overwriting annotations.yaml locally if existed
-func GenerateFunc(directory, outputDir, packageName, channels, channelDefault string, overwrite bool) error {
-	// clean the input so that we know the absolute paths of input directories
-	directory, err := filepath.Abs(directory)
-	if err != nil {
-		return err
-	}
-	if outputDir != "" {
-		outputDir, err = filepath.Abs(outputDir)
-		if err != nil {
-			return err
-		}
-	}
-
-	_, err = os.Stat(directory)
-	if os.IsNotExist(err) {
+func GenerateFunc(option ...BundleOption) error {
+	bundleConfig := &BundleConfig{}
+	bundleConfig.apply(option)
+	if err := bundleConfig.complete(); err != nil {
 		return err
 	}
 
 	// Determine mediaType
-	mediaType, err := GetMediaType(directory)
+	mediaType, err := GetMediaType(bundleConfig.bundleDir)
 	if err != nil {
 		return err
 	}
 
+	if mediaType == PlainType {
+		if err = synthesize.Generate(path.Join(bundleConfig.bundleDir, MetadataDir), ManifestsDir, bundleConfig.outputDir, bundleConfig.overwrite); err != nil {
```

Comment:
If I'm reading this correctly, the metadata directory is always `<given-bundle-dir>/metadata`, regardless of what an `annotations.yaml` file declares (if found), is that expected?

Reply:
I did not like this either, but there is no input for where else <metadata> could be. the <metadata> dir was hard coded to `metadata` by the dependency work. here we are saying that the metadata dir has tobe in the bundle dir. It is either this or we add another flag for there metadata dir is. It was previously not an issue because metadata was created either at the bundle level or  as a sub dir to the bundle.

Reply:
This may indicate an issue with the `generate` command's UX in general. On one hand it's generating a bundle from a set of k8s manifests which include a CSV, and on the other hand it's completing a valid bundle directory that's missing a CSV. Since we can't generate a CSV without a _mostly_ complete bundle, your suggestion may be the best way to discriminate between the two cases. 

Perhaps we can sidestep this problem altogether if we avoid writing out the CSV until `opm index add` is called.

Reply:
I want to lose our strict ruling of the bundle directory. I have pulled back the recursive walk to go through the entire bundle dir for any kube manifest that CSV uses, also getters non-inferable + dependencies.yaml. This does not tax cpu usage and ignores types and files we do not use.

Reply:
Please let me know if you agree this would be a better approach.

Reply:
divorced generate-csv and generate cmd. this is no longer a concern.